crates/flux-desugar/src/sort_resolver.rs

```diff
@@ -115,6 +123,10 @@ impl<'a> SortResolver<'a> {
             Ok(fhir::Sort::Bool)
         } else if ident.name == SORTS.real {
             Ok(fhir::Sort::Real)
+        } else if let Some(def_id) = self.parent_id
```

Comment:
Hmm, do you mean I should add a case `fhir::Sort::SelfParam` for the case when we are using `Self` as a TyParam?

Reply:
> Hmm, do you mean I should add a case `fhir::Sort::SelfParam` for the case when we are using `Self` as a TyParam?

More specifically, we should also be able to implement the trait. The following

```rust
impl MyTrait for i32 {
    #[flux::sig(fn<refine p: Self -> bool>(&Self{v: p(v)}) -> Self{v: p(v)})]
    fn foo1(&self) -> Self {
        *self
    }
}
```
is currently failing because `Self` is resolved to `fhir::Sort::SelfParam` with the def id of the impl block which doesn't make any sense.

Reply:
we could also not support `Self` in the impl. You have to write the sort explicitly 

```rust
impl MyTrait for i32 {
    #[flux::sig(fn<refine p: int -> bool>(&Self{v: p(v)}) -> Self{v: p(v)})]
    fn foo1(&self) -> Self {
        *self
    }
}
```

but then at some point we will need to be able to prove that `fn<refine p: int -> bool>(&Self{v: p(v)}) -> Self{v: p(v)})` is a valid refinement/specialization of `fn<refine p: Self -> bool>(&Self{v: p(v)}) -> Self{v: p(v)})`

Reply:
Do we currently support `Self` in the `impl`? (I can't recall...) 

The `refined_fn_in_trait.rs` supports it in the `trait` -- but not in the `impl` afaict?

I think ideally it would be nice to support it in `impl` so let me do this asap!

Reply:
yes, we do support `Self` in an `impl` when used as a type.

Reply:
Ok I have changed this to use the sort of `Self` appropriately for `impl` -- perhaps the code can be simplified more..